File: javascript/data-structures/Queue.js
const SinglyLinkedList = require("./SinglyLinkedList")

const Queue = class {
    count = 0;
    
    constructor() {
        this.data = new SinglyLinkedList();
        this.data.remove();
    }

    enqueue(val) {
        this.data.insert(val);
        this.count += 1;
        return this;
    }
    
    dequeue() {
        if (!this.isEmpty()) {
            const front = this.data.head.get();
            this.data.head = this.data.head.next;
            this.count -= 1;
            
            return front;
        } else {
            return "Queue is empty!";
        }
    }
    
    peek() {
        if (!this.isEmpty()) {
            return this.data.head.get();
        } else {
            return "Queue is empty!";
        }
    }

    isEmpty() {
        return this.count === 0;
    }

    toString() {
        return this.data.toString();
    }
}

module.exports = Queue;

File: java/src/main/java/io/github/oneminch/DataStructures/Queue.java
package io.github.oneminch.DataStructures;

import java.util.ArrayList;
import java.util.List;

public class Queue<T> {
    private final List<T> q;
    private int size;

    public Queue() {
        this.q = new ArrayList<T>();
        this.size = 0;
    }

    public void enqueue(T item) {
        this.q.add(item);
        this.size++;
    }

    public T dequeue() {
        if (isEmpty()) {
            return null;
        }

        this.size--;
        return this.q.remove(0);
    }

    public T peek() {
        return this.q.get(0);
    }

    public boolean isEmpty() {
        return this.size == 0;
    }

    public int getSize() {
        return this.size;
    }

    public List<T> getQueue() {
        return this.q;
    }
}
